Added new warning icon for missing data

v-aidaba opened the pull request.

v-aidaba committed:
Added new warning icon for missing data

Demonkratiy commented on stringResources/en-US/resources.resjson:
```diff
@@ -24,6 +24,10 @@
   "Visual_DeductThresholdDays": "Deduct threshold days",
   "Visual_DisplayUnits": "Display units",
+  "Visual_DataGap": "Data gaps",
+  "Visual_DataGapDescription": "Detect missing days in data",
```
let's mention that it works only if "Subtitle setting is on"

> "Detect missing days in data (works only if data Subtitle setting is on)"

v-aidaba replied:
Added as suggested

Demonkratiy commented on src/utils/dataGapDetector.ts:
```diff
@@ -0,0 +1,94 @@
+                gapDays += days > 1 ? days - 1 : 1;
+            } else if (currentGapStart) {
+                result.gaps.push({
+                    startDate: new Date(currentGapStart.getTime() + 86400000),
```
What does `86400000` mean? Why this value? Add comment to make it clear.

v-aidaba replied:
I replaced the hard-coded value 86400000 with a named constant:
private static  oneDayInMs = 24 * 60 * 60 * 1000 . Now it is fully clear that this value represents “one day in milliseconds,”

Demonkratiy commented on src/utils/dataGapDetector.ts:
```diff
@@ -0,0 +1,94 @@
+
+            const days = this.daysBetween(prev.x, curr.x);
+
+            if (!isPrevValid || !isCurrValid || days > 1) {
```
This `if` logic looks non-clear. Add comment in the beginning, explaining in short, what we check and what we do.

v-aidaba replied:
Added a clear comment above the if-logic explaining what we check:  invalid values or missing days between two dates.

Demonkratiy replied:
does "more than 1 day is missing between the dates" is required ? It means at least 2 days, or you want to tell "1 day or more are missing ..."?

v-aidaba replied:
The intention here is exactly “more than 1 day”, meaning 2 or more missing days between the dates.
A single missing day is treated as normal for time-series data (for example weekends or natural daily breaks), so I don’t mark it as a gap to avoid false-positive warnings.
Technically it’s possible to treat even a 1-day break as a gap, but I think that would generate unnecessary alerts.

v-aidaba commented on src/visualComponent/subtitleWarningComponent.ts:
```diff
@@ -203,4 +211,55 @@ export class SubtitleWarningComponent extends SubtitleComponent {
+        const hasGaps = dataRepresentation.dataGapInfo?.hasGaps ?? false;
+        const totalMissingDays = dataRepresentation.dataGapInfo?.totalMissingDays ?? 0;
+
+        let tooltipItems: VisualTooltipDataItem[] = [];
```
Changed let tooltipItems: VisualTooltipDataItem[] = [ ]; to let tooltipItems: VisualTooltipDataItem [ ]
Updated the  call to use tooltipItems || [ ] to handle the case where tooltipItems might be undefined

v-aidaba committed:
Added Data Gaps tests and updated PR changes

v-aidaba committed:
Merged main branch

Demonkratiy commented on CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 3.1.0.0
+* Add new option "Data Gaps" to detect missing days in time-series data and display a warning icon when gaps are found.
```
Please define changes type group (Fix, New feature, or something else)

v-aidaba replied:
Added

v-aidaba commented on src/utils/dataGapDetector.ts:
```diff
@@ -0,0 +1,103 @@
+            // - more than 1 day is missing between the dates
+            if (!isPrevValid || !isCurrValid || days > 1) {
+                if (!currentGapStart) currentGapStart = prev.x;
+                gapDays += days > 1 ? days - 1 : 1;  // If days are missing, count them; otherwise count this invalid point
```
I clarified the comment to clearly distinguish between: date gaps (days > 1 → count missing calendar days) invalid data points (NaN/null/Infinity → counted as 1 data issue, not a missing day) ․ The logic stays the same; only the explanation is now clearer.

Demonkratiy replied:
Consolidating: the underlying logic concern is still open and tracked in a more specific follow-up (the isolated invalid point is double-counted) — see #discussion_r3467447074. Resolving this thread to avoid duplication.

v-aidaba commented on src/settings/descriptors/dataGapDescriptor.ts:
```diff
@@ -0,0 +1,81 @@
+    });
+
+    public color: ColorPicker = new ColorPicker({
+        name: "color", 
```
Removed

v-aidaba committed:
PR changes

Demonkratiy commented on stringResources/en-US/resources.resjson:
```diff
@@ -80,5 +84,5 @@
   "Visual_Warning": "Warning",
   "Visual_YAxis": "Y-axis",
-  "Visual_ZeroLine": "Zero line"
+  "Visual_ZeroLine": "Zero line",
```
Trailing comma after the last entry makes this invalid JSON (and the trailing newline at EOF was removed). Please change `"Visual_ZeroLine": "Zero line",` back to `"Visual_ZeroLine": "Zero line"` and keep the newline at end of file.

v-aidaba replied:
Fixed.

Demonkratiy commented on styles/styles.less:
```diff
@@ -319,6 +319,18 @@
+            text-align: right;
+
+            &::before {
+                content: '\E70F';
```
`\E70F` is the same glyph already used by `.multiKpi_dataAge` (Stale Data) right above. The gap icon is visually indistinguishable from the stale-data icon, so users can't tell them apart. Please use a distinct warning glyph (e.g. `\E7BA`).

v-aidaba replied:
I switched the gap icon to a distinct glyph. Note: \E7BA isn't bundled in this visual's custom PowerVisuals font (it rendered as an empty block), so I used \E71B, a filled warning-triangle that exists in the font and is clearly distinguishable from the stale-data icon (\E70F).

Demonkratiy commented on src/utils/dataGapDetector.ts:
```diff
@@ -0,0 +1,103 @@
+            // - more than 1 day is missing between the dates
+            if (!isPrevValid || !isCurrValid || days > 1) {
+                if (!currentGapStart) currentGapStart = prev.x;
+                gapDays += days > 1 ? days - 1 : 1;  // If multiple days are missing, count the exact gap.If the date is present but the value is invalid (NaN/null), count it as 1 invalid data point.
```
A single invalid point between two consecutive valid days is counted twice: +1 when entering the invalid point and +1 when leaving it (via the `!isPrevValid` branch on the next iteration), so `valid → NaN → valid` yields `totalMissingDays = 2`. Consider counting an isolated invalid point only once. Note the current test only asserts `> 0`, so it doesn't catch this — worth adding an exact-count assertion.

v-aidaba replied:
Fixed. Removed the !isPrevValid branch from the gap condition so an isolated invalid point is counted only once,  plus a guard to count a leading invalid point once. I also tightened the test from toBeGreaterThan(0) to toBe(1) so it catches the regression.

Demonkratiy commented on src/utils/dataGapDetector.ts:
```diff
@@ -0,0 +1,103 @@
+
+    public static formatGapMessage(template: string, totalMissingDays: number): string {
+        return (template || `⚠️ ${totalMissingDays} missing days detected`)
+            .replace("${1}", totalMissingDays.toString());
```
`String.replace("${1}", ...)` replaces only the first occurrence. If a user's template contains `${1}` more than once, only the first is substituted. Please use a global replace (e.g. `replaceAll` or a regex `/\$\{1\}/g`).

v-aidaba replied:
Updated to a global replace

Demonkratiy commented on capabilities.json:
```diff
@@ -908,6 +908,38 @@
+            }
+          }
+        },
+        "backgroundColor": {
```
The analogous `staleData.background` is marked `"nullable": true` because its default is an empty string. `dataGap.backgroundColor` also defaults to `""` but isn't nullable, which can cause validation issues with an empty color. Consider adding `"nullable": true` here.

v-aidaba replied:
Added. It now mirrors staleData.background with nullable nested inside type.fill.solid.color

Demonkratiy commented on CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 3.1.0.0
+### New feature
+* Add new option "Data Gaps" to detect missing days in time-series data and display a warning icon when gaps are found.
```
The gap icon is shown by default (`isShown = true`), which is intentional and good — new users immediately notice gaps in their data. Please make this explicit in the changelog so users know the icon appears automatically whenever gaps are detected.

v-aidaba replied:
Done.

Demonkratiy reviewed (changes requested):
Thanks for the feature — the structure mirrors Stale Data nicely. A few things to address before merge: a JSON syntax error in the string resources, the gap icon currently being identical to the Stale Data icon, and a double-counting edge case in the gap detector. Details inline.

v-aidaba committed:
Fixed PR comments

Demonkratiy committed:
ci: modernize workflows and harden karma for ubuntu-latest

- Bump runner to ubuntu-latest and actions (checkout@v4, setup-node@v4)

- Update Node matrix to 20.x/22.x, add npm cache

- Reorder steps (ci before audit/outdated), drop dead eslint script

- Add concurrency with cancel-in-progress

- Add ChromeHeadlessCI launcher (--no-sandbox) for Ubuntu 24.04 sandbox

Demonkratiy committed:
chore(deps): refresh package-lock.json
